rfcs/20210518-prelu-postoperation/README.md

```diff
@@ -0,0 +1,55 @@
+# Introducing PReLU post operation support in oneDNN
+
+## Motivation
+
+This paper https://arxiv.org/abs/1502.01852 introduced PReLU activation
+function. The DIEN model uses PReLu as matmul activation.
+This gets expanded in TF graph.
+![](graph_prelu.png "Graph attention layer")
+
+It becomes a bottleneck for the Attention layer in DIEN. Currently oneDNN
+supports PReLu only as
+[standalone primitive](https://oneapi-src.github.io/oneDNN/dev_guide_prelu.html).
+
+## Proposal
+
+Extend [supported postops list](https://oneapi-src.github.io/oneDNN/dev_guide_attributes_post_ops.html)
+by adding PReLU forward inference postop.
+
+``` cpp
+    /// Appends a prelu forward post-op.
+    ///
+    /// The kind of this post-op is #dnnl::primitive::kind::prelu.
+    ///
+    /// The post-op can be defined as:
+    ///
+    ///      dst[:] <- prelu(dst[:], weights[:])
+    ///      prelu:
+    ///      dst[:] <- dst[:] if dst[:] > 0
+    ///      dst[:] <- dst[:] * weights[:] if dst[:] <= 0
+    ///
+    ///
+    /// @param weights_desc Learnable parameter memory descriptor.
+    void append_prelu_forward(const memory::desc &weights_desc) {
```

Comment:
Do we allow to pass a particular memory tag or it should be any?

Reply:
I think that the behaviour should follow standalone version - where we support "any tag" and encourage to use it. In case of jitted version, perfectly memory tag should be exactly the same as memory tag of destination tensor of the primitive it is being fused to. On the other hand in case of binary postops as far as I know we don't support "any" tag. In my opinion in case of prelu w should support it. What do you think ?

Reply:
Sorry for digging up a thread that is 1 month old. Here I am afraid that using mem::desc here would expose too many possible "unsupported" usages by the user. Hence my question is why not use something similar to output scale (so a mask and weights of shape x)? This is more restrictive, but should capture exactly the use cases for prelu (typically, weights are per-channel I believe).

Reply:
No problem. I was still waiting for feedback from fwk team. This approach is consistent with passing tensor in binary postops. Why do you think that there will be many unsupported cases ? Currently reference prelu kernel is supporting all types of weights broadcast. I think that we can reuse existing code in reference kernels, so there won't be any unsupported situations. Would I be able to use the "any" tag in your proposal? In case jit kernels for example convolution weights should have exactly the same format as output (case where user requests per tensor broadcast).

Reply:
The main difference is:
- the mask approach assumes that memory is in `x` layout, so there is no special path to support blocked layout, padding, ...
- the memory descriptor approach can describe a larger variety of layouts, and it is generally hard to support all layouts + broadcast combination in jit injector.

In general, unless we have a path to always dispatch the best implementation possible with the memory descriptor approach, I would advocate for the mask approach.

Reply:
Let's consider example. If prelu is post operation to convolution. Requested type of broadcast is per tensor (full broadcast). User set "any" tag as convolution destination (finally it can be blocked, plain etc). What memory layout should have weights tensor in that case ? In my opinion memory layout of weights should be exactly the same as convolution destination to generate most efficient and easy jit code without gather operations etc. If I understand correctly in mask approach the memory layout of the weights would be defined in advance e.g nchw ? Correct ?

Reply:
This case is a bit different than binary postop. In binary postop, the second argument generally comes from another layer, so it indeed can have different layout and have those advanced broadcast semantics.
For the Prelu postop, the second argument is a weight for the layer. Also, keep in mind post-ops target inference as of now.

Now on the shape of weights. From the [paper](https://arxiv.org/pdf/1502.01852.pdf), the [pytorch definition](https://pytorch.org/docs/stable/generated/torch.nn.PReLU.html), and the few topologies I know of that use prelu, I believe the dominant use case is to have 1 negative slope per channel (so layout is 1d and `C`). So the rigidity of the mask approach will likely be transparent for performance.

On the other hand, if we go with the md approach:
- if we support `any` and return blocked layouts, this actually could hurt performance by triggering extra reorders.
- if the layout is not `any` and the injector does not support it, we would have to fallback to reference code, which completely defeats the purpose of fusion as `ref convolution + fused prelu` will likely be much slower than `jitted convolution -> prelu primitive`. So from a user standpoint, fusing with postops could be slower than not fusing for those cases.

So my point is, sometimes it is better to have a more rigid API with predictable behavior, than to have a more general API for which only a few cases behave as expected (in that case, fused to most efficient impl).

Reply:
I agree that in case 1d and per oc broadcasts mask approach is better, because data format doesn't matter. But in my opinion it matters in other cases. Let's consider per tensor broadcast. Mask approach can possibly lead to performance degradation here. Let's go back to convolution example: let's say convolution output has nhwc format. Convolution before storing data has calculated ouptupt in vmm simd registers. Each of these registers contains a range of output channels. If data format of weights will be in that case nchw instead same as conv output it will have two side effects:

- instead of using offset to destination from convolution output we will in runtime calculate appropriate weights offset
- to apply postops on each register we will have to use gather operation, instead of reading continuous memory as if the weights layout was the same as the conv output layout

Reply:
I agree with your points. Note that this suboptimal case would equally apply for both options though.
However, when using mask, it seems that there are fewer such cases, and less potential for extra reorders.

Also, on the importance of this per-tensor case: how often do we need per tensor broadcast for prelu weights? To the best of my knowledge, prelu weights are almost always per channel.
And if we believe that this fusion will always be slower than when using a separate reorder, we could even restrict the weights to be either of shape {1} or shape {c}.